Dockerfile

```diff
@@ -1,61 +1,58 @@
-FROM debian:stretch-slim AS build
+FROM rust:slim AS build
 MAINTAINER Ivan <ivan@zderadicka.eu>
 
 ARG FEATURES
 
-RUN apt -o Acquire::https::No-Cache=True -o Acquire::http::No-Cache=True update &&\
-    apt-get install -y pkg-config openssl libssl-dev libtag1-dev libtagc0-dev curl yasm build-essential wget libbz2-dev zlib1g-dev &&\
-    curl -sL https://deb.nodesource.com/setup_8.x | bash - &&\
-    apt-get install -y nodejs 
-
-COPY . /audioserve_src
-
-RUN curl https://sh.rustup.rs -sSf | sh -s -- -y
-
-RUN mkdir ffmpeg-static &&\
-    curl -sL https://johnvansickle.com/ffmpeg/releases/ffmpeg-release-amd64-static.tar.xz | tar xJv -C ffmpeg-static --strip-components=1 --wildcards "*/ffmpeg" &&\
-    cp /ffmpeg-static/ffmpeg /usr/bin
+RUN apt-get update && \
+    apt-get install -y \
+    curl \
+    libbz2-dev \
+    libssl1.0-dev \
+    libtag1-dev \
+    libtagc0-dev \
+    openssl \
+    pkg-config \
+    yasm && \
+    mkdir ffmpeg-static && \
+    curl -sL https://johnvansickle.com/ffmpeg/releases/ffmpeg-release-amd64-static.tar.xz | \
+    tar xJv -C ffmpeg-static --strip-components=1 --wildcards "*/ffmpeg" && \
+    mv /ffmpeg-static/ffmpeg /usr/bin/ffmpeg
+
+WORKDIR /src
+COPY Cargo.toml Cargo.lock /src/
+# To cache this step, we only copy the toml file above
+RUN cargo fetch
+
+COPY . /src
+
+RUN cargo build --release ${FEATURES} && \
+    cargo test --release ${FEATURES}
 
+FROM node:8-alpine as frontend
 
-RUN export PATH=${HOME}/.cargo/bin:$PATH &&\
-    cd audioserve_src &&\
-    cargo build --release ${FEATURES} &&\
-    cargo test --release ${FEATURES}
+WORKDIR /src
+COPY client /src
 
-RUN cd audioserve_src/client &&\
-    npm install &&\
+RUN npm install && \
     npm run build
 
-RUN mkdir /ssl &&\
-    cd /ssl &&\
-    openssl req -newkey rsa:2048 -nodes -keyout key.pem -x509 -days 365 -out certificate.pem \
-        -subj "/C=CZ/ST=Prague/L=Prague/O=Ivan/CN=audioserve" &&\
-    openssl pkcs12 -inkey key.pem -in certificate.pem -export  -passout pass:mypass -out audioserve.p12 
-
```

Comment:
Although I agree with you comment, that it's not very useful for real deployment, I'd like to keep it here for testing purposes - it enables to test easily that ssl works - and users can easily add their own key and certificate later by mounting them to /audioserve/ssl ( or using rev. proxy). Maybe to make it non-default through env. vars?

Reply:
I was planning to add the mount command to the README instead, along with the certificate creation.

Since this is on the Dockerfile, it would be part of every docker image unless we move it to the shell script. So something like:

`docker run --env USE_INSECURE_CERT=1 izderadicka/audioserve:latest`

and generate the insecure cert inside the `audioserver.sh` if it was set?

Reply:
Then we also need openssl binary in final image.  I'd keep it in image but did not use it by default - let default be just plain HTTP as majority of user will use rev. proxy I think - just have an option to use bundled key for local TLS tests.